// localbeat-backend/models/CentralKitchenRawMaterial.js
const mongoose = require('mongoose');
const mongoosePaginate = require('mongoose-paginate-v2');

const CentralKitchenRawMaterialSchema = new mongoose.Schema({
  // From your sample data: SKU (10001, 10002, etc.)
  materialCode: {
    type: String,
    required: true,
    unique: true,
    trim: true,
    index: true
  },
  
  // From your sample data: Item Name (Bhujia, Bran Flakes, etc.)
  materialName: {
    type: String,
    required: true,
    trim: true,
    index: true
  },
  
  // From your sample data: Parent Category (always "Raw Materials")
  parentCategory: {
    type: String,
    required: true,
    default: 'Raw Materials',
    trim: true
  },
  
  // From your sample data: SubCategory Name (Bakery)
  subCategory: {
    type: String,
    required: true,
    trim: true,
    index: true
  },
  
  // From your sample data: Unit (kg, ltr)
  unitOfMeasure: {
    type: String,
    required: true,
    trim: true,
    enum: ['kg', 'ltr', 'g', 'ml', 'piece', 'box', 'pack'],
    default: 'kg'
  },
  
  // Additional fields for inventory management
  description: {
    type: String,
    trim: true,
    default: ''
  },
  
  // Pricing information
  unitPrice: {
    type: Number,
    required: true,
    min: 0,
    default: 0
  },
  
  // Stock management
  currentStock: {
    type: Number,
    required: true,
    min: 0,
    default: 0
  },
  
  minimumStock: {
    type: Number,
    required: true,
    min: 0,
    default: 10
  },
  
  maximumStock: {
    type: Number,
    required: true,
    min: 0,
    default: 1000
  },
  
  reorderPoint: {
    type: Number,
    required: true,
    min: 0,
    default: 20
  },
  
  // Supplier information
  supplierId: {
    type: String,
    trim: true,
    default: ''
  },
  
  supplierName: {
    type: String,
    trim: true,
    default: ''
  },
  
  // Storage requirements
  storageRequirements: {
    temperature: {
      type: String,
      trim: true,
      default: 'Room Temperature'
    },
    humidity: {
      type: String,
      trim: true,
      default: 'Normal'
    },
    specialConditions: {
      type: String,
      trim: true,
      default: ''
    }
  },
  
  // Shelf life in days
  shelfLife: {
    type: Number,
    min: 0,
    default: 365
  },
  
  // Status and tracking
  status: {
    type: String,
    enum: ['Active', 'Inactive', 'Discontinued'],
    default: 'Active',
    index: true
  },
  
  isActive: {
    type: Boolean,
    default: true,
    index: true
  },
  
  // Audit fields
  createdBy: {
    type: String,
    required: true,
    default: 'System'
  },
  
  updatedBy: {
    type: String,
    required: true,
    default: 'System'
  },
  
  // Optional Zoho sync tracking fields
  zohoItemId: {
    type: String,
    trim: true,
    default: '',
    sparse: true // Allow multiple null values
  },
  
  lastSyncedAt: {
    type: Date,
    default: null
  },
  
  zohoSyncStatus: {
    type: String,
    enum: ['', 'Synced', 'Updated', 'Failed'],
    default: ''
  }
}, {
  timestamps: true
});

// Add pagination plugin
CentralKitchenRawMaterialSchema.plugin(mongoosePaginate);

// Indexes for better performance
CentralKitchenRawMaterialSchema.index({ materialCode: 1, isActive: 1 });
CentralKitchenRawMaterialSchema.index({ subCategory: 1, status: 1 });
CentralKitchenRawMaterialSchema.index({ materialName: 'text', description: 'text' });

// Virtual for total value
CentralKitchenRawMaterialSchema.virtual('totalValue').get(function() {
  return this.currentStock * this.unitPrice;
});

// Pre-save middleware
CentralKitchenRawMaterialSchema.pre('save', function(next) {
  // Auto-generate material code if not provided
  if (!this.materialCode && this.materialName) {
    const prefix = this.subCategory.substring(0, 3).toUpperCase();
    const timestamp = Date.now().toString().slice(-4);
    this.materialCode = `${prefix}-${timestamp}`;
  }
  
  next();
});

// Static methods
CentralKitchenRawMaterialSchema.statics.findByCategory = function(category) {
  return this.find({ subCategory: category, isActive: true });
};

CentralKitchenRawMaterialSchema.statics.findLowStock = function() {
  return this.find({ 
    currentStock: { $lte: { $expr: '$reorderPoint' } },
    isActive: true 
  });
};

module.exports = (connection) => connection.model('CentralKitchenRawMaterial', CentralKitchenRawMaterialSchema);
